### ISSUE TITLE
Is the matched bracket popup themable?

### ISSUE THREAD

**Issue by Docteh** [bobbylight/RSyntaxTextArea#371]
This might be a really silly question. Are the matched bracket popups controlled by the theme?

https://github.com/bobbylight/RSyntaxTextArea/blob/3dc9a584928fa721a6fd431332b9019938cb48f7/RSyntaxTextArea/src/main/java/org/fife/ui/rsyntaxtextarea/MatchedBracketPopup.java#L63

I don't know if TipUtil is part of RSTA or part of java

Starting point: 
https://github.com/konrad91/OneDarkArduino/issues/5

Ended up skimming through a chunk of the Arduino IDE's code trying to figure out what the pop up is called. I was about to give up until I started reading closed issues here and found https://github.com/bobbylight/RSyntaxTextArea/issues/175

Under the hood the Arduino IDE appears to be using either 3.0.2 or 3.0.3. 

I'll come back to this in a bit, just suspending the search before I go chasing after TipUtil

**Comment by Docteh** [comment]
TipUtil is part of RSTA
https://github.com/bobbylight/RSyntaxTextArea/blob/21306ca2e457652eb270e8db260487d547d10c88/RSyntaxTextArea/src/main/java/org/fife/ui/rsyntaxtextarea/focusabletip/TipUtil.java#L75-L96

I'm using the windows platform, and I'm not sure of a proper way to change the default styling in a sensible way. 

But for amusement: https://www.formdev.com/flatlaf/themes/ it released a version 1.0 just last week.

I downloaded the flatlaf library, and made a `.bat` file

```batch

set EXEDIR=%cd%
set CLASSPATH=%EXEDIR%\flatlaf-1.0.jar;%EXEDIR%\lib\apple.jar;%EXEDIR%\lib\arduino-core.jar;%EXEDIR%\lib\batik-1.8.jar;%EXEDIR%\lib\batik-anim-1.8.jar;%EXEDIR%\lib\batik-awt-util-1.8.jar;%EXEDIR%\lib\batik-bridge-1.8.jar;%EXEDIR%\lib\batik-codec-1.8.jar;%EXEDIR%\lib\batik-css-1.8.jar;%EXEDIR%\lib\batik-dom-1.8.jar;%EXEDIR%\lib\batik-ext-1.8.jar;%EXEDIR%\lib\batik-gvt-1.8.jar;%EXEDIR%\lib\batik-parser-1.8.jar;%EXEDIR%\lib\batik-rasterizer-1.8.jar;%EXEDIR%\lib\batik-script-1.8.jar;%EXEDIR%\lib\batik-squiggle-1.8.jar;%EXEDIR%\lib\batik-svg-dom-1.8.jar;%EXEDIR%\lib\batik-svgpp-1.8.jar;%EXEDIR%\lib\batik-transcoder-1.8.jar;%EXEDIR%\lib\batik-util-1.8.jar;%EXEDIR%\lib\batik-xml-1.8.jar;%EXEDIR%\lib\bcpg-jdk15on-152.jar;%EXEDIR%\lib\bcprov-jdk15on-152.jar;%EXEDIR%\lib\commons-codec-1.7.jar;%EXEDIR%\lib\commons-compress-1.8.jar;%EXEDIR%\lib\commons-exec-1.1.jar;%EXEDIR%\lib\commons-httpclient-3.1.jar;%EXEDIR%\lib\commons-lang3-3.8.1.jar;%EXEDIR%\lib\commons-logging-1.0.4.jar;%EXEDIR%\lib\commons-net-3.3.jar;%EXEDIR%\lib\commons-io-2.6.jar;%EXEDIR%\lib\jackson-annotations-2.9.5.jar;%EXEDIR%\lib\jackson-core-2.9.5.jar;%EXEDIR%\lib\jackson-databind-2.9.5.jar;%EXEDIR%\lib\java-semver-0.8.0.jar;%EXEDIR%\lib\jmdns-3.5.5.jar;%EXEDIR%\lib\jtouchbar-1.0.0.jar;%EXEDIR%\lib\log4j-api-2.12.0.jar;%EXEDIR%\lib\log4j-core-2.12.0.jar;%EXEDIR%\lib\slf4j-simple-1.7.22.jar;%EXEDIR%\lib\slf4j-api-1.7.22.jar;%EXEDIR%\lib\jna-4.2.2.jar;%EXEDIR%\lib\jna-platform-4.2.2.jar;%EXEDIR%\lib\jsch-0.1.50.jar;%EXEDIR%\lib\jssc-2.8.0-arduino4.jar;%EXEDIR%\lib\pde.jar;%EXEDIR%\lib\rsyntaxtextarea-3.0.3-SNAPSHOT.jar;%EXEDIR%\lib\xml-apis-1.3.04.jar;%EXEDIR%\lib\xml-apis-ext-1.3.04.jar;%EXEDIR%\lib\xmlgraphics-commons-2.0.jar
java\bin\java -Dswing.systemlaf=com.formdev.flatlaf.FlatDarkLaf processing.app.Base
```
Normal:
![image](https://user-images.githubusercontent.com/190571/109741472-fb2d9c00-7b81-11eb-9644-461d098c379c.png)
With flatlaf:
![image](https://user-images.githubusercontent.com/190571/109739702-de439980-7b7e-11eb-8b3d-95c5e599277d.png)


So in summary, Is the matched bracket popup themable? Not yet. It looks like the background and the border are two elements that are only pulled from javas UIManager, while the rest can be overwritten.

**Comment by bobbylight** [comment]
So this isn't part of the RSTA `Theme` yet, but I have added a system property that should help here.

The default behavior is still the same - that is, matched bracket popups render like tool tips.  The idea here is most applications are themed in such a way that light LookAndFeels are paired with light RSTA `Theme` instances, and dark LookAndFeels with dark RSTA `Theme`s.  In such cases, this default behavior always looks good (a la the FlatLaF screenshot above).

However, there may be cases where this doesn't hold true.  For example, in apps that allow the user to select a LookAndFeel separately from an RSTA theme - this would allow e.g. a light application overall but a dark RSTA.  In such cases the default behavior doesn't work, as  you'll end up with light text in a light popup, or dark text in a dark popup, e.g. your "Normal" screenshot above.

In those cases, you can now set this system property:

```bash
-Drsta.matchedBracket.considerTextAreaBackground=true
```

This can be set programmatically in an application via:

```java
System.setProperty(MatchedBracketPopup.PROPERTY_CONSIDER_TEXTAREA_BACKGROUND, "true");
```

When this is set to `true`, matched bracket popups will sniff out whether the parent RSTA instance is using a white background or not.  If it is, it renders like a tool tip (it makes the assumption that white editors are always in light apps - might not be guaranteed, but who in their right mind would drop a light editor in a dark-themed app?).  If it is *not* white, it renders itself with the text area's background color instead of the tool tip background color, and a complimentary border.  This ensures that the line of code in this popup always has good contrast, since it should look identical to the actual editor contents.

I think this is a fair tradeoff for now, since I'd imagine most apps either have one fixed LookAndFeel/RSTA Theme combo, or a light/dark mode toggle, and in either case they'd control the entirety of the app's colors and can thus ensure these popups look good.  But if not, this system property covers all bases.

The `RSyntaxTextAreaDemoApp` submodule of the project has been updated to set this system property, along with a comment explaining its use.  The demo was also expanded to allow changing both the LookAndFeel and the RSTA Theme at runtime so you can play with various combinations and view the changes in the matched bracket popup.

Let me know if this works for you, at least for a short-term fix.  It's currently available in Sonatype as 3.1.3-SNAPSHOT.

**Comment by bobbylight** [comment]
Closing as a workaround is now in place via the system property described above.
